README.md
# langtons_ant
First assignment for Com S 327 at [ISU](https://www.iastate.edu)

This program creates a Langton's ant Simulation and renders the frames into an
mpeg video.

## Run this program

This project includes a Makefile, so to build the project:
```bash
$ make
```
For more information on `make`, click [here](https://www.gnu.org/software/make/)

This will create an binary called `langtons_ant`. This can be 
run with the default settings using:
```bash
$ ./langtons_ant
```

To view the different modes available for the ant, show the
help menu:
```bash
$ ./langtons_ant -h
```

#### Run Modes

This project includes several different modes that affect the ant's collision 
with the borders.

- **Normal**   - The ant will stop when it reaches any border.
- **Cylinder** - The ant will wrap around on the x-axis borders, but will stop on the y-axis borders.
- **Torus**    - The ant will wrap around on both axes' borders. 
- **Reflect**  - The ant will reflect off of all borders when it reaches them

To change the run mode, pass in a flag at runtime.

| Run Mode | Flag                                            |
|:--------:|:----------------------------------------------- |
| Normal   | This is the default mode - _No flag necessary_. |
| Cylinder | use `-c` or `--cylinder`.                       |
| Torus    | use `-t` or `--torus`.                          |
| Reflect  | use `-r` or `--reflect`.                        |

> You can pass in multiple run modes, however only one will be selected.
> Torus will override Cylinder, and Reflect will override both Cylinder and Torus.

#### Variable Canvas Size

The length or width of the canvas used can be changed based on flags passed in:

| Dimension | Flag                                            |
|:---------:|:----------------------------------------------- |
| Length    | use `-l<int>` or `--length <int>`.              |
| Width     | use `-w<int>` or `--width <int>`.               |

> Both the length and width have a default value of 128 if no flag is passed in.
> The range of these argument is [1,255]. Additionally, for encoding to work 
> properly, the area of the canvas (length * width) must be divisible by 8.

```bash
$ ./langtons_ant -w64 --length 120
```

#### Variable Frame Skip Rate

The number of frames to skip can be varied using a flag.  This will speed up or slow down
the final .mpg video based on if the number is high or low.  To vary the skip rate, pass in 
the following flag:

```
$ ./langtons_ant -s45
$ ./langtons_ant --skip 42
```

After the program completes, a file called `langton.mpg` should be 
available for you to view!

> This program does use [NetPBM](http://netpbm.sourceforge.net) to translate 
> the ppm frames to an mpeg video.  This must be installed for the mpeg video 
> to be created.


CHANGELOG.md
# Change Log
All notable changes to this project (langtons_ant) will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## [1.3.0] - 2016-8-29

### Added
- Variable canvas size - you can now specify a width and length using a flag
- Variable frame skip rate - you can now specify the frame skip rate with a flag

### Fixed
- Bug where cylinder would not terminate when ant crossed y-axis top border
- Typo in README.md

## [1.2.0] - 2016-8-28

### Added 
- Parser for different options
- Cylinder Mode - ant wraps left to right
- Torus Mode - ant wraps on all axes
- Reflect Mode - ant reflects when it hits the border

## [1.1.0] - 2016-08-27

### Changed
- Moved Ant specific code to separate folder `ant/`
- Moved Encoding specific code to separate folder `encode/`
- Updated Makefile with new build instruction rules

## [1.0.0] - 2016-08-25

### Added
- Makefile for building the project.

### Changed
- Implemented basic ant path.


parser/parser.c
//
//  parser.c
//  langtons_ant
//
//  Created by Srinivas Dhanwada on 8/28/16.
//  Copyright © 2016 dhanwada. All rights reserved.
//

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include "parser.h"

// Default board length
#define BOARD_LENGTH_DEFAULT 128

// Default skip length
#define SKIP_LENGTH_DEFAULT 50

// Defined from parser.h
int BORDER_MODE = 0;
int ARG_LENGTH = BOARD_LENGTH_DEFAULT;
int ARG_WIDTH = BOARD_LENGTH_DEFAULT;
int ARG_SKIP = SKIP_LENGTH_DEFAULT;

const char* help_text = "Usage: langtons_ant [options]\n\n"
                        "-c,      --cylinder     Run the ant in cylinder mode -- ant will wrap across the\n"
                        "                        x-axis.\n"
                        "-h,      --help         Print this help message.\n"
                        "-l<int>, --length <int> The width of the canvas.  The default value is 128.\n"
                        "                        This value can be in the following range: [1,255].\n"
                        "-r,      --reflect      Run the ant in reflect mode -- ant will change\n"
                        "                        direction when it hits a border.  This will override\n"
                        "                        both cylinder and torus modes!\n"
                        "-s<int>, --skip <int>   The number of frames to skip when creating the mpg\n"
                        "                        video. This value must be greater than 0.\n"
                        "-t,      --torus        Run the ant in torus mode -- ant will wrap across both\n"
                        "                        the x and y axes.  This will override cylinder mode!\n"
                        "-w<int>, --width <int>  The width of the canvas.  The default value is 128.\n"
                        "                        This value can be in the following range: [1,255].\n";

int is_number(char* str) {
    int val = atoi(str);
    return (val > 0 || (val == 0 && strlen(str) == 1)) ? val : -1;
}

void parse_args(int argc, char** argv) {
    
    if(argc == 1) {
        return;
    }
    
    int flag;
    int help_flag = 0;
    int number = -1;
    
    opterr = 0;
    
    while(1) {
        // Setup options
        static struct option long_options[] = {
            {"cylinder", no_argument, 0, 'c'},
            {"help", no_argument, 0, 'h'},
            {"length", required_argument, 0, 'l'},
            {"reflect", no_argument, 0, 'r'},
            {"skip", required_argument, 0, 's'},
            {"torus", no_argument, 0, 't'},
            {"width", required_argument, 0, 'w'},
            {0, 0, 0, 0}
        };
        
        int option_index = 0;
        flag = getopt_long(argc, argv, "chrtl:s:w:", long_options, &option_index);
        
        // flags are finished, break out of while loop
        if(flag == -1) {
            break;
        }
        
        // parse which flag was found
        switch (flag) {
            case 0:
                if(long_options[option_index].flag != 0) {
                    break;
                }
                printf("option: %s", long_options[option_index].name);
                if(optarg) {
                    printf(" with args %s", optarg);
                }
                printf("\n");
                break;
                
            case 'c':
                if(BORDER_MODE == 0) {
                    BORDER_MODE = 1;
                }
                break;
                
            case 'h':
                help_flag = 1;
                break;
                
            case 'l':
                number = is_number(optarg);
                if(number <= 0 || number >= 256) {
                    fprintf(stderr, "Error: length must be an integer in range [1, 255].\n");
                    exit_with_help();
                }
                ARG_LENGTH = number;
                break;
                
            case 'r':
                BORDER_MODE = 3;
                break;
                
            case 's':
                number = is_number(optarg);
                if(number <= 0) {
                    fprintf(stderr, "Error: skip must be a positive integer.\n");
                    exit_with_help();
                }
                ARG_SKIP = number;
                break;
                
            case 't':
                if(BORDER_MODE < 2) {
                    BORDER_MODE = 2;
                }
                break;
                
            case 'w':
                number = is_number(optarg);
                if(number <= 0 || number >= 256) {
                    fprintf(stderr, "Error: width must be an integer in range [1, 255].\n");
                    exit_with_help();
                }
                ARG_WIDTH = number;
                break;
                
            case '?':
                opterr = 1;
                break;
                
            default:
                exit_with_help();
        }
    }
    
    if(opterr) {
        fprintf(stderr, "Error: invalid flag parsed.\n");
        exit_with_help();
    }
    
    if(help_flag) {
        exit_with_help();
    }
}

void exit_with_help() {
    printf("%s", help_text);
    exit(0);
}
